### DISCUSSION TITLE
ProvideOpusFrame Panics, doesn't provide error

### DISCUSSION THREAD

**Discussion by tmaffia** [disgoorg/disgo#318] [Help] [answered]
**Describe the bug**
When using OpusReader, I get a panic when reading an Opus file. I've tested the opus file and it plays in VLC, i've no reason to suspect its invalid.

I've no experience streaming audio so this entire thing is a bit of a mystery to me, otherwise i'd try to solve this and send a PR. I'm sort of stumbling in the dark. I see some functions included in the voice package that seem to be there to help with streaming audio over UDP, i'm attempting use them. Though, I noticed the examples, and the documentation both do not use these functions. I'm guessing they just haven't been updated. I'm sure there is more a chance that i've made an obvious mistake.

**Error**
The error which gets logged if there is any:
```
panic: runtime error: slice bounds out of range [:1399285583] with length 3840
```

**To Reproduce**
```go
func writeAudio(conn voice.Conn) error {
	file, err := os.Open("../../audio/file.opus")
	if err != nil {
		return err
	}

	r := voice.NewOpusReader(file)
	b, err := r.ProvideOpusFrame()
	for err != nil {
		conn.UDP().Write(b)
	}
	r.Close()

	return nil
}
```

The code panics at r.ProvideOpusFrame()

**Expected behavior**
I would expect the function to return an error, and give me some idea what is going wrong.

**Disgo Version(please complete the following information):**
 - Version v0.16.9

**Comment by tmaffia** [comment]
Unrelated:
I noticed the voice example in this repo uses a .dca file, and works as is. I tried to circumvent this problem by converting my audio file to .dca with two different libraries I've found. Neither worked, they appear to be abandoned. Is there an easy way to create a .dca audio file?

**Comment by topi314** [comment, marked as answer]
`conn.UDP().Write()` expects you to write exactly one Opus frame.

This means you need to split the frames from an opus frame yourself

The example uses the `voice.NewOpusReader` which implements the `OpusFrameProvider` interface

You need to call `conn.SetOpusFrameProvider()` with the `voice.NewOpusReader()` to make it play the whole opus file

`voice.NewOpusReader()` expects the following format of the opus file:
Frame length (4 bytes) Opus Frame (x Bytes)
And repeat 
You can check the implementation here  https://github.com/disgoorg/disgo/blob/master/voice/opus.go#L20
